pre-commit.py

```diff
@@ -0,0 +1,271 @@
+                        ANSWER_START_INDICATOR,
+                        nb_file
+                    )
+                )
+
+            unterminated_start = False
+
+        else:
+            if not unterminated_start:
+                question_source.append(line)
+
+    if unterminated_start:
+        raise Exception(
+            "{} without corresponding {} in {}"
+            .format(
+                ANSWER_START_INDICATOR,
+                ANSWER_END_INDICATOR,
+                nb_file
+            )
+        )
+
+    return question_source
+
+
+def generate_notebooks(exercise_nb_path):
+    """
+    Takes in the exercise notebook and generates
+    the question and answer notebooks
+    """
+
+    with open(exercise_nb_path) as exercise_nb_file:
+        exercise_nb = json.load(exercise_nb_file)
+
+        # copy the notebooks
+        questions = copy.deepcopy(exercise_nb)
+        answers = copy.deepcopy(exercise_nb)
+
+        # generate the new cells for the questions
+        # and answers notebooks
+        questions["cells"] = []
+        answers["cells"] = []
+
+        for cell in exercise_nb["cells"]:
+            source = cell["source"]
+
+            question_cell = copy.copy(cell)
+            question_cell["source"] = generate_question_source(
+                exercise_nb_path,
+                source
+            )
+            # ignore the outputs cells for questions
+            question_cell["outputs"] = []
+            questions["cells"].append(question_cell)
+
+            answer_cell = copy.copy(cell)
+            answer_cell["source"] = generate_answer_cell(source)
+            answers["cells"].append(answer_cell)
+
+        return questions, answers
+
+
+def create_directories(nb_file):
+    """
+    Creates the directories to store the
+    nbs if they don't exist
+    """
+    notebook_dir = os.path.dirname(nb_file)
+    questions_dir = os.path.join(notebook_dir, "questions")
+    answers_dir = os.path.join(notebook_dir, "answers")
+
+    if not os.path.exists(questions_dir):
+        os.makedirs(questions_dir)
+
+    if not os.path.exists(answers_dir):
+        os.makedirs(answers_dir)
+
+    return questions_dir, answers_dir
+
+
+def write_notebook(nb, nb_dir):
+    """
+    Writes the notebook to the corresponding
+    directory
+    """
+    nb_path = os.path.join(
+        nb_dir,
+        "notebook.ipynb"
+    )
+
+    with open(nb_path, "w") as nb_file:
+        json.dump(nb, nb_file)
+
+
+def get_nb_path(nb_dir):
+    """
+    Helper function to get the notebook
+    path from the directory
+    """
+    return os.path.join(
+        nb_dir,
+        "notebook.ipynb"
+    )
+
+
+def generate_preview(nb_dir):
+    """
+    Converts the notebooks to html files
+    then moves the files to give them the correct name
+    """
+    nb_path = get_nb_path(nb_dir)
+
+    # create the corresponding preview
+    command = "jupyter nbconvert --to html {} --output {}".format(
```

Comment:
Have we tested that the previews that his generates render correctly in QF? We customize the html that's produced in research when a notebook is shared and when someone uses the "download as html" option, so I wouldn't expect the preview to render correctly without those customizations in place.  In the past I think we've used the "download as html" option to generate these previews.

Here's the custom template file that gets used by nbconvert in research: https://github.com/quantopian/research_env/blob/53cb330dac43b314086aef7d0eec3a884860970c/user_notebook/config/templates/nbconvert.html

Here's where we tell the notebook server in research to use that template file:
https://github.com/quantopian/research_env/blob/master/base/src/shared/shared/config_utils.py#L88

Reply:
I didn't know! This should be easy to incorporate as you can pass in the notebook template as a keyword argument. I'm actually going to be moving this into the lectures_internal repo and will make the changes there @TimShawver